File: JavaScript/Problem001SumMultiples.js
// This script solves Project Euler Problem 1: Sum of multiples of two numbers below a limit.
// It interacts with an HTML page where users can input two numbers and a limit.
// The core calculation uses the Inclusion-Exclusion Principle to find the sum of multiples.
// For example, finding the sum of multiples of 3 or 5 below 1000.

/**
 * Calculates the sum of all numbers strictly below `limit` that are multiples
 * of a single number `divisibleBy`.
 *
 * @param {number} divisibleBy - The number whose multiples are to be summed. Must be a positive integer.
 * @param {number} limit - The exclusive upper bound for the multiples. Must be a positive integer.
 * @returns {number} The sum of the multiples.
 * @throws {Error} If `divisibleBy` or `limit` are not positive integers.
 */
function sumMultiplesOf(divisibleBy, limit) {
  // Validate inputs: must be positive integers.
  if (
    !Number.isInteger(divisibleBy) ||
    !Number.isInteger(limit) ||
    divisibleBy <= 0 ||
    limit <= 0
  ) {
    throw new Error(
      "Inputs to sumMultiplesOf must be positive integers."
    );
  }
  let sum = 0;
  // Iterate from 0 up to (but not including) limit.
  for (let i = 0; i < limit; i++) {
    // If 'i' is a multiple of 'divisibleBy', add it to the sum.
    if (i % divisibleBy === 0) {
      sum += i;
    }
  }
  return sum;
}

/**
 * Calculates the sum of multiples of `divisor1` OR `divisor2` strictly below `limit`.
 * This function uses the Inclusion-Exclusion Principle: Sum(A or B) = Sum(A) + Sum(B) - Sum(A and B).
 * Sum(A and B) is implemented by finding multiples of `divisor1 * divisor2`. This is accurate
 * if `divisor1` and `divisor2` are coprime (e.g., 3 and 5). For non-coprime numbers,
 * the LCM should ideally be used, but `divisor1 * divisor2` is standard for this specific problem's interpretation.
 *
 * @param {number} divisor1 - The first divisor. Must be a positive integer.
 * @param {number} divisor2 - The second divisor. Must be a positive integer.
 * @param {number} limit - The exclusive upper bound for the multiples. Must be a positive integer.
 * @returns {number} The total sum of multiples of `divisor1` or `divisor2` below `limit`.
 * @throws {Error} If `divisor1`, `divisor2`, or `limit` are not positive integers, or if any input is not an integer.
 */
function calculateSumOfMultiplesLogic(divisor1, divisor2, limit) {
  // Validate that all inputs are integers.
  if (
    !Number.isInteger(divisor1) ||
    !Number.isInteger(divisor2) ||
    !Number.isInteger(limit)
  ) {
    throw new Error("All inputs must be integers.");
  }
  // Validate that all inputs are positive.
  if (divisor1 <= 0 || divisor2 <= 0 || limit <= 0) {
    throw new Error("All inputs must be positive integers.");
  }

  // Sum of multiples of the first divisor.
  const sumDivisor1 = sumMultiplesOf(divisor1, limit);
  // Sum of multiples of the second divisor.
  const sumDivisor2 = sumMultiplesOf(divisor2, limit);
  // Sum of common multiples (multiples of divisor1 * divisor2).
  // This addresses the double-counting in sumDivisor1 and sumDivisor2.
  const sumCommonMultiples = sumMultiplesOf(divisor1 * divisor2, limit);

  // Apply the Inclusion-Exclusion Principle.
  return sumDivisor1 + sumDivisor2 - sumCommonMultiples;
}

/**
 * Event handler for the "calculate" button click.
 * It orchestrates fetching inputs, calling the calculation logic,
 * and updating the DOM with the results or errors.
 */
function calculateSimply() {
  // Get DOM elements for displaying results and execution time.
  const resultElement = document.getElementById("result01");
  const executionTimeElement = document.getElementById("executionTime01");
  
  // Clear previous results and execution time from the DOM.
  resultElement.textContent = "";
  executionTimeElement.textContent = "";

  // Start performance timer to measure execution duration.
  const startTime = performance.now();

  try {
    // Fetch input values from the respective DOM input fields.
    // Parse them as base-10 integers.
    const multiple1_1 = parseInt(
      document.getElementById("multiple1_1").value,
      10
    );
    const multiple1_2 = parseInt(
      document.getElementById("multiple1_2").value,
      10
    );
    const limit1 = parseInt(document.getElementById("limit1").value, 10);

    // Initial validation: Check if parsing resulted in NaN (Not-a-Number).
    // This catches empty fields or non-numeric text before calling the core logic.
    if (isNaN(multiple1_1) || isNaN(multiple1_2) || isNaN(limit1)) {
      throw new Error("Inputs must be valid numbers. Please check your entries.");
    }
    
    // Call the core calculation logic with the parsed input values.
    const sumAllMultiples = calculateSumOfMultiplesLogic(
      multiple1_1,
      multiple1_2,
      limit1
    );

    // If calculation is successful, display the formatted sum in the result DOM element.
    resultElement.textContent = `${sumAllMultiples.toLocaleString()}`;
  } catch (error) {
    // If any error occurs (either from input parsing/validation or core logic),
    // display the error message in the result DOM element.
    resultElement.textContent = error.message; 
  } finally {
    // This block executes regardless of whether an error occurred or not.
    // End performance timer.
    const endTime = performance.now();
    // Calculate execution time in milliseconds, formatted to two decimal places.
    const executionTime = (endTime - startTime).toFixed(2);
    // Display the execution time in the designated DOM element.
    executionTimeElement.textContent = `${executionTime} ms`;
  }
}

// Ensures that the DOM is fully loaded and parsed before attaching event listeners.
document.addEventListener("DOMContentLoaded", function () {
  // Get the "Calculate" button element.
  const calculateButton = document.getElementById("calculateSimplyButton");
  // Attach the 'calculateSimply' function to the button's 'click' event.
  if (calculateButton) {
    calculateButton.addEventListener("click", calculateSimply);
  } else {
    console.error("Calculate button not found."); // Optional: error handling if button is missing
  }
});

// Export functions for testing if module is available (e.g., in Node.js/Jest environment)
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { sumMultiplesOf, calculateSumOfMultiplesLogic };
}

File: Python/Problem001SumMultiples.py
# Solves Project Euler Problem 1: Sum of multiples of two numbers below a limit.
# This script provides a Flask web interface for the calculation.
# The core logic uses the Inclusion-Exclusion Principle.
# For example, to find the sum of multiples of 3 or 5 below 1000.

from flask import Flask, render_template, request, jsonify
import time

# Initialize the Flask application
app = Flask(__name__)

# Function to calculate the sum of multiples
def sumMultiples(divisibleBy: int, limit: int) -> int:
    """
    Calculates the sum of all numbers strictly below `limit` that are multiples
    of a single number `divisibleBy`.

    This function uses a direct iteration approach.

    Parameters:
    divisibleBy (int): The number whose multiples are to be summed. Must be positive.
    limit (int): The exclusive upper bound for the multiples. Must be positive.

    Returns:
    int: The sum of multiples of `divisibleBy` below `limit`.

    Raises:
    ValueError: If `divisibleBy` or `limit` are not positive integers.
    """
    if not isinstance(divisibleBy, int) or not isinstance(limit, int) or divisibleBy <= 0 or limit <= 0:
        raise ValueError("Inputs to sumMultiples must be positive integers.")
    return sum(i for i in range(limit) if i % divisibleBy == 0)

# Core logic function
def calculate_sum_of_multiples_for_two_divisors(divisor1: int, divisor2: int, limit: int) -> int:
    """
    Calculates the sum of all numbers strictly below `limit` that are multiples
    of `divisor1` OR `divisor2`.

    This function employs the Inclusion-Exclusion Principle:
    Sum(A or B) = Sum(A) + Sum(B) - Sum(A and B)

    Where:
    - Sum(A) is the sum of multiples of `divisor1`.
    - Sum(B) is the sum of multiples of `divisor2`.
    - Sum(A and B) is the sum of multiples of numbers that are multiples of BOTH
      `divisor1` and `divisor2`. For this problem, this is typically taken as
      multiples of `divisor1 * divisor2`. This is accurate if `divisor1` and
      `divisor2` are coprime (like 3 and 5). If they are not coprime, the
      Least Common Multiple (LCM) should be used for `Sum(A and B)`; however,
      Project Euler Problem 1 usually implies `divisor1*divisor2` for the common term.

    Parameters:
    divisor1 (int): The first number whose multiples are considered. Must be positive.
    divisor2 (int): The second number whose multiples are considered. Must be positive.
    limit (int): The exclusive upper bound for the multiples. Must be positive.

    Returns:
    int: The sum of multiples of `divisor1` or `divisor2` below `limit`.

    Raises:
    ValueError: If `divisor1`, `divisor2`, or `limit` are not positive integers,
                or if any input is not an integer.
    """
    if not all(isinstance(arg, int) for arg in [divisor1, divisor2, limit]):
        raise ValueError("All inputs must be integers.")
    if not (divisor1 > 0 and divisor2 > 0 and limit > 0):
        raise ValueError("All inputs must be positive integers.")

    # Calculate sum of multiples for the first divisor
    sum_divisor1 = sumMultiples(divisor1, limit)
    # Calculate sum of multiples for the second divisor
    sum_divisor2 = sumMultiples(divisor2, limit)
    # Calculate sum of multiples for numbers divisible by both (common multiples)
    # Using divisor1 * divisor2 as the common multiple term, common for this problem.
    sum_common_multiples = sumMultiples(divisor1 * divisor2, limit)

    # Apply Inclusion-Exclusion Principle
    return sum_divisor1 + sum_divisor2 - sum_common_multiples

@app.route("/", methods=["GET", "POST"])
def index():
    """
    Handles web requests for the sum of multiples calculator.
    - GET requests: Displays the input form.
    - POST requests: Processes form data, calculates the sum, and displays the result.
    """
    result = None
    execution_time = None
    # Dictionary to store form values to repopulate the form upon submission or error
    form_values = {} 

    if request.method == "POST":
        try:
            # Extract form data from the POST request and convert to integers
            multiple1_1 = int(request.form["multiple1_1"])
            multiple1_2 = int(request.form["multiple1_2"])
            limit1 = int(request.form["limit1"])
            
            # Store current form values for repopulation
            form_values = {
                "multiple1_1": multiple1_1,
                "multiple1_2": multiple1_2,
                "limit1": limit1
            }

            # Start timer for execution time measurement
            start_time = time.time()

            # Call the core logic function to calculate the sum of multiples
            # This is the main calculation step.
            result = calculate_sum_of_multiples_for_two_divisors(multiple1_1, multiple1_2, limit1)

            # End timer and calculate execution time in milliseconds
            end_time = time.time()
            execution_time = round((end_time - start_time) * 1000, 2)

        except ValueError as e:
            # Handle ValueError if inputs are invalid (e.g., non-positive, non-integer)
            # The error message from the core logic functions will be displayed.
            result = str(e)
            # Ensure form values are captured even if initial int conversion failed or values were empty
            if not form_values or not all(form_values.values()): # Repopulate if dict is empty or has falsey values from successful int() but error later
                 form_values = {
                    "multiple1_1": request.form.get("multiple1_1", ""), # Use .get to avoid KeyError
                    "multiple1_2": request.form.get("multiple1_2", ""),
                    "limit1": request.form.get("limit1", "")
                }
        except Exception as e: # Catch any other unexpected errors during processing
            result = f"An unexpected error occurred: {str(e)}"
            # Ensure form values are captured for repopulation
            if not form_values or not all(form_values.values()):
                 form_values = {
                    "multiple1_1": request.form.get("multiple1_1", ""),
                    "multiple1_2": request.form.get("multiple1_2", ""),
                    "limit1": request.form.get("limit1", "")
                }

    # Render the HTML template, passing the result, execution time, and form values
    # These variables are used by Jinja2 in the 'index.html' template.
    return render_template("index.html", result=result, execution_time=execution_time, form_values=form_values)


if __name__ == "__main__":
    # Runs the Flask development server
    # Debug mode is on, which provides helpful error messages and auto-reloads on code changes.
    app.run(debug=True)
